src/ot1d.jl

```diff
@@ -0,0 +1,283 @@
+where ``h`` is a convex function.
+μ: 1-D distribution (e.g. `μ = Distributions.Normal(0,1)`)\n
+ν: 1-D distribution (e.g. `μ = Distributions.Normal(0,1)`)\n
+
+Returns the Optimal Transport Plan as a function
+  ```math
+  T(x)=F_\\nu^{-1}(F_\\mu(x)).
+  ```
+"""
+function otPlan1d(c,μ::Distributions.UnivariateDistribution,ν::Distributions.UnivariateDistribution)
+  T(x) = Distributions.quantile(ν,Distributions.cdf(μ,x))
+  return T
+end
+
+
+"""
+otCostPlan1d(c,μ::Distributions.UnivariateDistribution,
+            ν::Distributions.UnivariateDistribution)`
+
+Calculates the Optimal Transport Cost between μ to ν, where
+they are 1-Dimensional distributions and the cost
+function is of the form ``c(x,y) = h(|x-y|)`` such that
+``h`` is a convex function.
+
+Parameters:\n
+c: The cost function, which should be of the form ``c(x,y) = h(abs(x-y))``
+where ``h`` is a convex function.
+μ: 1-D distribution (e.g. `μ = Distributions.Normal(0,1)`)\n
+ν: 1-D distribution (e.g. `μ = Distributions.Normal(0,1)`)\n
+
+Returns the Optimal Transport Cost and the Plan as a function
+  ```math
+  T(x)=F_\\nu^{-1}(F_\\mu(x)).
+  ```
+"""
+function otCostPlan1d(c,μ::Distributions.UnivariateDistribution,ν::Distributions.UnivariateDistribution)
+    g(μ,ν,x) = c(quantile(μ,x),quantile(ν,x))
+    f(x) = g(μ,ν,x)
+    T(x) = Distributions.quantile(ν,Distributions.cdf(μ,x))
+    return quadgk(f,0,1)[1], T
+end
+
+"""
+otCost1d(c,u::Vector,u_weights::Vector,v::Vector,v_weights::Vector)
+otCost1d(c,μ::Distributions.DiscreteNonParametric,ν::Distributions.DiscreteNonParametric)
+
+Calculates the Optimal Transport Cost between μ to ν, where
+they are 1-Dimensional distributions and the cost
+function is of the form ``c(x,y) = h(|x-y|)`` such that
+``h`` is a convex function.
+
+For the first case we have parameters:\n
+c: The cost function, which should be of the form ``c(x,y) = h(abs(x-y))``
+where ``h`` is a convex function. \n
+u        : Vector of point mass location    (e.g. `u = collect(1:10)`)\n
+u_weights: Vector of weights for each point (e.g. `u_weights = ones(10)*0.1`)\n
+v        : Vector of point mass location    (e.g. `v = collect(10:20)`)\n
+v_weights: Vector of weights for each point (e.g. `v_weights = ones(10)*0.1`)\n
+
+
+For the second case we have parameters:\n
+c: The cost function, which should be of the form ``c(x,y) = h(abs(x-y))``
+where ``h`` is a convex function. \n
+μ: Finite Discrete Distribution (e.g. `μ = DiscreteNonParametric(u, u_n)`)\n
+ν: Finite Discrete Distribution (e.g. `ν = DiscreteNonParametric(v, v_n)`)\n
+"""
+function otCost1d(c,u::Vector,u_weights::Vector,v::Vector,v_weights::Vector)
+
+  @assert length(u) == length(u_weights)
+  @assert length(v) == length(v_weights)
+  @assert sum(u_weights) ≈ 1.0  atol=1e-7
+  @assert sum(v_weights) ≈ 1.0  atol=1e-7
+
+  cost = 0
+  perm_u = sortperm(u)
+  perm_v = sortperm(v)
+  u_sorted = u[perm_u]
+  v_sorted = v[perm_v]
+  u_w_sorted = u_weights[perm_u]
+  v_w_sorted = v_weights[perm_v]
```

Comment:
This can be removed if only `DiscreteNonParametric` is accepted:

```suggestion
  perm_u = sortperm(u)
  perm_v = sortperm(v)
  u_sorted = u[perm_u]
  v_sorted = v[perm_v]
  u_w_sorted = u_weights[perm_u]
  v_w_sorted = v_weights[perm_v]
```